### PR TITLE
fix(context): publish a restored UE context only once it is built

### PR DESCRIPTION
**Pull request by midwell**
## The defect

`DbFetch` stored the UE into `UePool` and its `RanUe` into `RanUePool`, and *then* wrote the event
channel, the four per-UE loggers and the instance identity. Everything after the publication is
state a reader needs, so a lookup landing in between got a context with no loggers and no identity.

Those readers do not hold `dbMutex`: `AmfUeFindBySupi` and `AmfUeFindByGuti` reach `UePool`
directly, so the reader is a service request arriving while the restore is still running.

It also became easier to reach recently. Since 826, a request naming a UE with no event channel
runs its handler inline instead of failing on the nil channel — so it now gets as far as the
procedure, and `SmContextStatusNotifyProcedure` and its siblings write through `ue.ProducerLog`,
which is one of the fields the restore is still assigning.

This is the reorder that 802 considered and did not take: its own comment above these lines says
*"the UE is published to UePool three lines down, so from that point the entry can be replaced by
another goroutine while this function is still using it"*, and it worked around that by reading the
`RanUe` once rather than by moving the publish. That comment is updated here, because after the
move it was describing a distance that no longer exists.

## The fix

Move both `Store` calls to the end of the function. Nothing between the two positions reads either
pool — that is what makes the move safe, and it is the whole argument for the change: a reader now
either does not find the UE, or finds it complete.

## What this does not close

The safety argument above covers `DbFetch` itself and the unlocked readers. It does not cover every
writer. `NewAmfUe` publishes a live registration through `AddAmfUeToUePool` (`context/context.go:282`)
under `amfContextMutex`, not `dbMutex`, so nothing serialises it against this publish. Delaying the
publish plausibly **widens** the window in which a live registration's fresher entry lands first and
is then overwritten by the later, staler restored one.

That is a pre-existing gap — two writers to one pool under two different locks — and it is the same
class the change already defers to `upstream-amf-sbi-ue-state-races`. Stated rather than absorbed
into "safe", because a reader of this diff deserves to know the argument is about readers, not
about every writer.

`context/transaction.go`'s comment justifying the package-logger fallback cites two writers of
`TxLog`: `AttachRanUe` under `ue.Mutex`, and `DbFetch` after publication. This PR closes the second
of those. The first still stands, so the fallback's choice of logger is still right — only one of
its two cited reasons is gone. Left for a follow-up rather than touched here, to keep this to one
defect.

## What the test does and does not do

It is **not** a regression test for the ordering, and should not be read as one. It passes under
the old order too — I reverted the reorder and ran it to check, rather than assuming — because by
the time `DbFetch` returns the context is complete either way. That window has no hook to drive it from a unit test, and a probabilistic test
for it would be worth less than the ordering argument in the code.

What it pins:

- the post-condition a reader depends on, read back **through the pool** rather than from the
  return value, since the pool is how every other goroutine reaches this UE;
- that `DbFetch` needs nothing from the datastore but the single document read — the stub embeds a
  nil `DBInterface`, so the other twenty-one methods are promoted from nil and any call panics.

Both are what make publishing last safe, and both would catch a later change that put a pool read
back in between.

It is also the first test to drive `DbFetch` itself; the package had none.

## Gate

`pre-commit run --all-files` clean, with the three Go hooks run in the pinned `linux/amd64`
container: `golangci-lint run` (v2.13.2) and `staticcheck ./...` (v0.8.1) clean, and
`go test -race ./... -count=1` clean across every package.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: context/db_fetch_publish_test.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	}
+
+	ranUe := pooled.RanUe[models.ACCESSTYPE__3_GPP_ACCESS]
+	if ranUe == nil {
+		t.Fatal("restored UE has no 3GPP RanUe")
+		return
+	}
+	if _, ok := AMF_Self().RanUePool.Load(ranUe.AmfUeNgapId); !ok {
+		t.Fatal("restored RanUe was never published into RanUePool")
+	}
```
[the remark this answers is not in the document]
Comment (midwell) [reply context/db_fetch_publish_test.go]: Fixed in e92a4ed — you are right, and chasing it turned up something worse underneath.

The presence check could not tell whether DbFetch published anything, exactly as you say: NewRanUe already stores a RanUe under that AmfUeNgapId (context/amf_ran.go:152). It compares identity now.

What I did not expect is that removing the Store *did* fail the test, which looked like a counterexample to your point. Reading the failure line rather than trusting it: NewAmfRanId leaves AnType empty, AttachRanUe keys ue.RanUe by AnType, so the fixture RanUe landed under the empty access type and ue.RanUe[ACCESSTYPE__3_GPP_ACCESS] was nil. MarshalJSON then wrote amfUeNgapId 0, and the assertion only passed because nothing else uses key 0. The document never described a 3GPP-attached UE at all — so the fixture was wrong in a way your comment did not name, and your comment is what found it.

Both are fixed. With AnType set and the assertion comparing identity, removing the Store now fails with "RanUePool still holds the fixture's RanUe, not the restored one".

File: context/db_fetch_publish_test.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+// checked, not assumed -- because by the time DbFetch returns the context is complete either
+// way. It is not a regression test for the ordering: that window has no hook to drive it from
+// a unit test, and a probabilistic one would be worth less than the argument in the code,
+// which is that nothing between the two positions reads either pool.
+//
+// What it does pin is the post-condition a reader depends on, expressed through the pool
+// rather than the return value, and -- via the embedded nil DBInterface -- that DbFetch needs
+// nothing from the datastore but the one read. Both are what make publishing last safe, and
+// both would catch a later change that reintroduced a pool read between building and
+// publishing.
```
[the remark this answers is not in the document]
Comment (midwell) [reply context/db_fetch_publish_test.go]: Correct, and withdrawn in e92a4ed.

I had conflated two different things. The embedded nil DBInterface does catch a later change reaching for another *datastore* method — any of the other twenty-one panics. Nothing in this test catches a *pool* read inserted between building and publishing, because such a read has no effect the test can observe. Writing "both would catch a later change that reintroduced a pool read" took the true half and extended it over the false half.

The claim is gone from the test comment and the PR description, and the new commit message withdraws it explicitly rather than quietly — the first commit message still carries it and I cannot amend that one, since the branch was rebased by the maintainer and force-pushing would discard that work. The squash will carry both messages, so the correction needs to be visible in the second.

No instrumentation added in its place: I would rather the test claim less than have it claim coverage it does not have.

**Review by gab-arrobo — APPROVED** [review]
